src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder_ImplementingTypes.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.FindSymbols
+{
+    using SymbolSet = HashSet<INamedTypeSymbol>;
+
+    internal static partial class DependentTypeFinder
+    {
+        /// <summary>
+        /// Implementation of <see cref="SymbolFinder.FindImplementationsAsync(ISymbol, Solution, IImmutableSet{Project}, CancellationToken)"/> for 
+        /// <see cref="INamedTypeSymbol"/>s
+        /// </summary>
+        public static Task<ImmutableArray<INamedTypeSymbol>> FindAndCacheImplementingTypesAsync(
+            INamedTypeSymbol type,
+            Solution solution,
+            IImmutableSet<Project> projects,
+            bool transitive,
+            CancellationToken cancellationToken)
+        {
+            return FindTypesFromCacheOrComputeAsync(
+                type, solution, projects,
+                transitive ? s_typeToTransitivelyImplementingTypesMap : s_typeToImmediatelyImplementingTypesMap,
+                c => FindWithoutCachingImplementingTypesAsync(type, solution, projects, transitive, c),
+                cancellationToken);
+        }
+
+        private static async Task<ImmutableArray<INamedTypeSymbol>> FindWithoutCachingImplementingTypesAsync(
+            INamedTypeSymbol type,
+            Solution solution,
+            IImmutableSet<Project> projects,
+            bool transitive,
+            CancellationToken cancellationToken)
+        {
+            // Only an interface can be implemented.
+            if (type?.TypeKind == TypeKind.Interface)
+            {
+                static bool typeMatches(SymbolSet s, INamedTypeSymbol t, bool transitive)
+                    => TypeHasBaseTypeInSet(s, t, transitive) || TypeHasInterfaceInSet(s, t, transitive);
```

Comment:
hrmm... i think that bit may not be necessary at all.  renaming these def helped call that out as weird.  i will investigate and remove if not needed.

Reply:
ok.  so this is necessary.  i will doc in code.  The reason for this is a simple scenario like:

`interface IFoo { } class Base : IFoo { } class Derived : Base { }`

Both `Base` and `Derived` are implementations of IFoo.  So, when searching, we'll first add IFoo to the set, then later on add Base.  Then when examining `Derived` we want it to match because it's `.BaseType` is in that set.

One thing to help see that is the following code in that method:

```c#
                var allTypes = await FindTypesAsync(type, solution, projects,
                    typeMatches: typeMatches,
                    shouldContinueSearching: s_isInterfaceOrNonSealedClass,
```

Notice how we keep searching if we hit an interface that matches or a *non-sealed* class.  If the class was sealed it couldn't have any more derived classes off of it, so there's no point looking for those impls.  But for non-sealed classes we have to keep going.

Reply:
Note: we have tests for that.